tests/php/Control/Middleware/HTTPCacheControlMiddlewareTest.php

```diff
@@ -68,6 +68,65 @@ public function testSetMaxAge($state, $immutable)
         }
     }
 
+    public function testEnableCacheWithMaxAge()
```

Comment:
I think these tests aren't needed and we should just update the provider and the existing tests...

Reply:
For this particular test (`testEnableCacheWithMaxAge`), I agree with you - it's already covered by `testSetMaxAge`.

But otherwise, I find this particular use of PHPUnit providers quite unreadable actually - it took me a while to figure out what the meaning of `$immutable` is in this context. I think providers are fine when they contain one dimension, since that's something you can keep in your head. But otherwise, DRY principles apply differently to unit tests: In my opinion, clarity (and sometimes verbosity) is more important than brevity. I can understand what `testEnableCacheWithMaxAgeAppliesWhenLevelDoesNot` tests. I couldn't understand this if it was mixed in as a third array element in `provideCacheStates`, even when using [named data provider sets](https://phpunit.de/manual/6.5/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers)

Reply:
Update: The difference with `testEnableCacheWithMaxAge` (compared to `testSetMaxAge`) is the max-age setting via argument rather than setter. Not all methods in your data provider have this argument, because it doesn't make sense for all of them - so you'd either have to use a different data provider, or add more confusing dimensions to the existing one. And since the assertions change between `enableCache` and `publicCache` usage, I don't think a new data provider capturing both cases will be more readable.